src/lib/transition_frontier_persistence/transition_frontier_persistence.ml

```diff
@@ -106,19 +53,180 @@ end = struct
+
+  (* We would like the operation for moving a root to be atomic to avoid phantom
+     write and having our persistent data to be left in a weird state. To do this, we
+     first write our data new root data into a temp file. Then, we add the best
+     tip and remove old transitions in batch manner in transition_storage.
+     Afterwards, we requickly rename our file of the new root_data as the old
+     file *)
+  let atomic_move_root
+      ( {transition_storage; root_snarked_ledger= _; root_storage; logger= _}
+      as t )
+      {Frontier_diff.best_tip; removed_transitions; new_root; new_scan_state} =
+    let open Deferred.Or_error.Let_syntax in
+    let temp_location = root_storage.path ^ ".temp" in
+    let%bind old_root, old_scan_state = read_root t in
+    let%bind () = lift_or_error @@ set_root t new_root new_scan_state in
+    Deferred.return
+      (let open Result.Let_syntax in
+      let%map response =
+        Transition_storage.Batch.with_batch transition_storage ~f:(fun batch ->
+            let%map parent =
+              apply_add_transition t best_tip
+                ~write:(Batch.set_transition batch)
+            in
+            let removed_transitions =
+              List.map removed_transitions ~f:(fun state_hash ->
+                  let removed_transition =
+                    read_transition t state_hash |> Option.value_exn
+                  in
+                  let body_hash = consensus_state removed_transition in
+                  Batch.remove_transition batch state_hash ;
+                  body_hash )
+            in
+            {Diff_mutant.parent; removed_transitions; old_root; old_scan_state}
+        )
+      in
+      (* HACK: We would like to have this synchronous and not context switch to
+          another task by the async scheduler because we would like the original
+          file of the root data to updated as quickly as possible. Renaming a
+          file should be very fast so it's okay to wait for a bit *)
+      Core.Sys.rename temp_location root_storage.path ;
+      response)
+
+  let apply_diff (type mutant) t (diff : mutant Diff_mutant.t) :
+      mutant Deferred.Or_error.t =
+    match diff with
+    | Move_root request -> atomic_move_root t request
+    | Add_transition
+        (transition_with_hash :
+          (External_transition.Verified.t, State_hash.t) With_hash.t) ->
+        Deferred.return
+        @@ apply_add_transition t transition_with_hash
+             ~write:(set_transition t)
+
+  let handle_diff (t : t) acc_hash (frontier_diff : Frontier_diff.t) =
+    let open Deferred.Or_error.Let_syntax in
+    let compute_hash diff_mutant =
+      let%map mutant = apply_diff t diff_mutant in
+      Diff_mutant.hash acc_hash diff_mutant mutant
+    in
+    match frontier_diff with
+    | Add_transition transition ->
+        compute_hash (Diff_mutant.Add_transition transition)
+    | Move_root move_root -> compute_hash (Diff_mutant.Move_root move_root)
```

Comment:
This only computes the hash of the mutant. You need to compute the hash of the previous hash, the diff content, and the mutant all together. Checkout the RFC for more details on the incremental hash specification.